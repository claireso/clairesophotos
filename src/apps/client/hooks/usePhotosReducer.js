import { useReducer } from 'react'

import { getPhotos } from '@client/api'
import photosReducer, { TYPES } from './reducer/photos'

export default (initialState = {}) => {
  const [state, dispatch] = useReducer(photosReducer, initialState)

  return [
    state,
    {
      setLoading: isLoading =>
        dispatch({ type: TYPES.SET_LOADING, isLoading: isLoading }),
      getPhotos: async page => {
        try {
          const response = await getPhotos(page)

          dispatch({ type: TYPES.GET_PHOTOS_SUCCESS, response })
        } catch (err) {
          throw new Error('Can not get photos')
        }
      }
    }
  ]
}
