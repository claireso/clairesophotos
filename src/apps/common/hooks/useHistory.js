import { useEffect } from 'react'

export default ({ onPopstate } = {}) => {
  useEffect(() => {
    onPopstate && window.addEventListener('popstate', onPopstate)

    return () => {
      onPopstate && window.removeEventListener('popstate', onPopstate)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const pushState = (state, title, url) => {
    window.history.pushState(state, title, url)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }

  return {
    pushState
  }
}
