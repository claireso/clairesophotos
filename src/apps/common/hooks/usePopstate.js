import { useEffect } from 'react'

export default fn => {
  useEffect(() => {
    // listen history
    window.addEventListener('popstate', fn)

    return () => {
      window.removeEventListener('popstate', fn)
    }
  }, [fn])
}
