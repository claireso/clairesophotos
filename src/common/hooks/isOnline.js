import { useState, useEffect } from 'react'

export default () => {
  const [isOnline, setIsOnline] = useState(true)

  const handleOnlineStatus = () => setIsOnline(true)

  const handleOfflineStatus = () => setIsOnline(false)

  useEffect(() => {
    setIsOnline(navigator.onLine)

    window.addEventListener('offline', handleOfflineStatus)
    window.addEventListener('online', handleOnlineStatus)

    return () => {
      window.removeEventListener('offline', handleOfflineStatus)
      window.removeEventListener('online', handleOnlineStatus)
    }
  }, [])

  return isOnline
}