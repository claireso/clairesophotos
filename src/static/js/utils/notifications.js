import urlBase64ToUint8Array from './urlBase64ToUint8Array'

export default {
  areDenied() {
    return Notification.permission === 'denied'
  },

  areGranted() {
    return Notification.permission === 'granted'
  },

  areDefault() {
    return Notification.permission === 'default'
  },

  async subscribe(registration) {
    const pushPublicKey = await this.getPushPublicKey()

    if (!registration) {
      registration = await this.getRegistration()
    }

    if (!pushPublicKey) return

    const convertedPushPublicKey = urlBase64ToUint8Array(pushPublicKey)

    const subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: convertedPushPublicKey
    })

    await fetch('/subscriptions', {
      method: 'post',
      headers: {
        'Content-type': 'application/json'
      },
      body: JSON.stringify({
        subscription: subscription
      })
    })
  },

  async getRegistration() {
    try {
      return await navigator.serviceWorker.ready
    } catch (err) {
      throw new Error('Banner: can not get registration')
    }
  },

  async getPushPublicKey() {
    try {
      const response = await fetch('/push-public-key')
      return await response.text()
    } catch (err) {
      throw new Error('Banner can not get public key')
    }
  }
}
