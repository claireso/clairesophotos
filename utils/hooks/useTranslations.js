import React from 'react'
import PropTypes from 'prop-types'

const TranslationsContext = React.createContext({})

const TranslationsProvider = ({ children, translations }) => (
  <TranslationsContext.Provider value={translations}>
    {children}
  </TranslationsContext.Provider>
)

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

const useTranslations = () => React.useContext(TranslationsContext)

export { TranslationsProvider, useTranslations }
