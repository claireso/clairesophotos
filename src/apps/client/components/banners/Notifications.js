import React, { useState, useEffect, useContext } from 'react'
import styled from 'styled-components'

import TranslationsContext from '@common/context/Translations'
import notifications from '@common/utils/notifications'

import Flash from '../Flash'

const ButtonSubscribe = styled.a.attrs(() => ({
  href: '#',
  role: 'button'
}))`
  color: inherit;
  text-decoration: none;

  &:before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }

  &:hover {
    border-bottom: 1px solid currentColor;
  }
`

/**
 * checkSubscription: check if subscription is expired
 * @return Boolean
 */
const checkSubscription = async () => {
  try {
    const subscription = await notifications.getSubscription()

    // expired subscription
    if (!subscription) {
      return true
    }

    return false
  } catch (err) {
    throw new Error('Banner: Can not check subscription')
  }
}

export default () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useContext(TranslationsContext)

  const subscribe = async event => {
    event && event.preventDefault()

    try {
      await notifications.subscribe()
      hideBanner()
    } catch (err) {
      // user decline
      if (notifications.areDenied()) {
        hideBanner()
        return
      }

      throw new Error('Banner: can not subscribe')
    }
  }

  const hideBanner = () => setIsVisible(false)
  const showBanner = () => setIsVisible(true)

  useEffect(() => {
    // do not display banner if service worker not support
    // do not display banner if push not enabled
    // do not display banner in safari
    if (
      !('serviceWorker' in navigator) ||
      (process.env.NODE_ENV !== 'production' &&
        process.env.NODE_ENV !== 'test') ||
      !process.env.IS_PUSH_ENABLED ||
      window.safari
    ) {
      return
    }

    // show banner if user
    // - has not already subscribed
    // - or has not denied notification
    if (notifications.areDefault()) {
      showBanner()
      return
    }

    // check if subscription is expired
    if (notifications.areGranted()) {
      checkSubscription()
        .then(isSubscriptionExpired => {
          if (isSubscriptionExpired) {
            subscribe()
          }
        })
        .catch(() => {})
    }
  }, [])

  if (!isVisible) return null

  return (
    <Flash onClose={hideBanner}>
      <ButtonSubscribe onClick={subscribe}>
        {translations.bannerNotifications}
      </ButtonSubscribe>
    </Flash>
  )
}
