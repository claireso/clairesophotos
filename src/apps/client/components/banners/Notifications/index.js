import React, { useState, useEffect, useContext, useCallback } from 'react'
import styled from 'styled-components'

import TranslationsContext from '@common/context/Translations'
import notifications from '@common/utils/notifications'

import Flash from '../../Flash'

const ButtonSubscribe = styled.a.attrs(() => ({
  href: '#',
  role: 'button'
}))`
  color: inherit;
  text-decoration: none;

  &:before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }

  &:hover {
    border-bottom: 1px solid currentColor;
  }
`

export default () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useContext(TranslationsContext)

  const subscribe = useCallback(async event => {
    event && event.preventDefault()

    try {
      await notifications.subscribe()
      hideBanner()
    } catch (err) {
      // user decline
      if (notifications.areDenied()) {
        hideBanner()
        return
      }

      throw new Error('Banner: can not subscribe')
    }
  }, [])

  const hideBanner = useCallback(() => setIsVisible(false), [])
  const showBanner = useCallback(() => setIsVisible(true), [])

  useEffect(() => {
    // do not display banner if service worker or Notification is not support
    // do not display banner if push is not enabled
    // do not display banner in safari
    if (
      !process.env.IS_PUSH_ENABLED ||
      !('serviceWorker' in navigator) ||
      !('Notification' in window) ||
      (process.env.NODE_ENV !== 'production' &&
        process.env.NODE_ENV !== 'test') ||
      window.safari
    ) {
      return
    }

    // do not display banner if if user has denied notifications
    if (notifications.areDenied()) {
      return
    }

    // check if user has already subscribed
    notifications
      .getSubscription()
      .then(subscription => {
        if (!subscription) {
          showBanner()
        }
      })
      .catch(err => {
        throw new Error('Banner can not get subscription')
      })
  }, [])

  if (!isVisible) return null

  return (
    <Flash onClose={hideBanner}>
      <ButtonSubscribe onClick={subscribe}>
        {translations.bannerNotifications}
      </ButtonSubscribe>
    </Flash>
  )
}
