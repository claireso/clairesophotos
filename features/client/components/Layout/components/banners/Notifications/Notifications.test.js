import { render, fireEvent, waitFor } from '@testing-library/react'

import { TranslationsProvider } from '@utils/hooks/useTranslations'
import * as notifications from '@services/notifications'

import Notifications from './index'

describe('<Notifications />', () => {
  const renderComponent = () =>
    render(
      <TranslationsProvider translations={global.__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsProvider>
    )

  beforeEach(() => {
    global.setNotificationPermission()
  })

  test('should render component', (done) => {
    const spy = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve(null))

    const { container } = renderComponent()

    setImmediate(() => {
      expect(container).toMatchSnapshot()
      spy.mockRestore()
      done()
    })
  })

  test('should not render component (user has already subscribed)', (done) => {
    const spy = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve({}))

    const { container } = renderComponent()

    setImmediate(() => {
      expect(container).toMatchSnapshot()
      spy.mockRestore()
      done()
    })
  })

  test('should not render component (user has blocked notifications)', () => {
    global.setNotificationPermission('denied')

    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
    global.setNotificationPermission()
  })

  test('should subscribe and hide banner', async (done) => {
    // mock func subscribe
    const spy = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.resolve())

    // mock func getSubscription
    const spySubscription = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve(null))

    const { container, getByText } = renderComponent()

    await waitFor(() => getByText(/^Enable notifications/i), {
      container
    })

    fireEvent.click(getByText(/^Enable notifications/i))

    setImmediate(() => {
      expect(container).toMatchSnapshot()
      spy.mockRestore()
      spySubscription.mockRestore()
      done()
    })
  })

  test('should not subscribe and hide banner (denied story)', async (done) => {
    // mock func subscribe
    const spy = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.reject())

    // mock func getSubscription
    const spySubscription = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve(null))

    const { container, getByText } = renderComponent()

    await waitFor(() => getByText(/^Enable notifications/i), {
      container
    })

    global.setNotificationPermission('denied')

    fireEvent.click(getByText(/^Enable notifications/i))

    setImmediate(() => {
      expect(container).toMatchSnapshot()
      spy.mockRestore()
      spySubscription.mockRestore()
      done()
    })
  })
})
