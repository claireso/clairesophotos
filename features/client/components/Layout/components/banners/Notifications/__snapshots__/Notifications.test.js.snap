// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> should not render component (user has already subscribed) 1`] = `<div />`;

exports[`<Notifications /> should not render component (user has blocked notifications) 1`] = `<div />`;

exports[`<Notifications /> should not subscribe and hide banner (denied story) 1`] = `<div />`;

exports[`<Notifications /> should render component 1`] = `
.c2 {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.c2:hover {
  border-bottom: 1px solid currentColor;
}

.c1 {
  background: var(--flash-default);
  color: var(--flash-default-text);
  font-size: 1.4rem;
  margin: 0 0 var(--gutter);
  padding: 2rem;
  position: relative;
  text-align: center;
}

.c1 + .c0 {
  border-top: 3px solid var(--flash-default-border);
}

.c4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  outline: none;
  padding: 0;
  margin: 0;
  border-radius: 0.2rem;
  font-size: 1.2rem;
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  padding: 0 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
}

.c4 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c4 + .c3 {
  margin: 0 0 0 0.8rem;
}

.c4 > svg {
  margin: 0;
}

<div>
  <div
    class="c0 c1"
    style="opacity: 0;"
  >
    <a
      class="c2"
      href="#"
      role="button"
    >
      Enable notifications to be alerted of new publication
    </a>
    <button
      class="Buttonsstyles__DefaultButton-jpizpn-0 Buttonsstyles__StyledButton-jpizpn-1 c3 c4"
    >
      <svg
        height="16"
        viewBox="0 0 22 28"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.281 20.656c0 0.391-0.156 0.781-0.438 1.062l-2.125 2.125c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-4.594-4.594-4.594 4.594c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-2.125-2.125c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l4.594-4.594-4.594-4.594c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l2.125-2.125c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l4.594 4.594 4.594-4.594c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l2.125 2.125c0.281 0.281 0.438 0.672 0.438 1.062s-0.156 0.781-0.438 1.062l-4.594 4.594 4.594 4.594c0.281 0.281 0.438 0.672 0.438 1.062z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`<Notifications /> should subscribe and hide banner 1`] = `<div />`;
